Prevent mismatched braces in format string

chall8908 opened the pull request:
Not sure if this is the best way to go about this.  It works, but just counting braces seems a bit... brute force, I suppose.

Fixes #37

burke commented:
This reactivates an old nagging thought for me, that we could implement this at a lower level as a more legitimate parser, and not this pile of hacks. It would be a really fun little project, and if done correctly, could allow us to bind to it from any language.

I think this is probably a sensible idea but I'm kind of undecided about how permissive vs. strict we'd like to be with input strings here.

burke commented:
@lavoiesl, @lugray, any thoughts on this? Should we be doing this validation or no?

chall8908 commented:
@lugray I've fixed the issues with and added additional tests for out-of-order, leading, and trailing literal closing braces.

So, those particular instances should no longer occur.  Additionally, this means that closing braces are emitted whenever they don't appear to match to opening braces.  This jives with what I would expect the formatter to do in those instances, but it _is_ a change from current behavior.

lavoiesl commented:
Thanks for the patch!
